File: solutions/haskell/bob/1/src/Bob.hs
module Bob (responseFor) where

-- import qualified Data.Set as Set
import qualified Data.Text as T
import           Data.Text (Text)
import qualified Data.Char as Char

-- ideally should just need Data.Text, and write way more idiomatic code

responseFor :: Text -> Text
responseFor xs
  | isYellQ    = T.pack "Calm down, I know what I'm doing!"
  | isQuestion = T.pack "Sure."
  | isYell     = T.pack "Whoa, chill out!"
  | isSilence  = T.pack "Fine. Be that way!"
  | otherwise  = T.pack "Whatever."
    where
      isQuestion = T.isSuffixOf (T.singleton '?') (T.strip xs)
      isYell = ((T.toUpper xs) == xs && (or (map Char.isLetter (T.unpack xs))))
      isSilence = (T.words xs) == [] -- could use strip too?
      isYellQ = (T.isSuffixOf (T.singleton '?') xs) && ((T.toUpper xs) == xs && (or (map Char.isLetter (T.unpack xs))))

File: solutions/cpp/bob/1/bob.cpp
#include "bob.h"

namespace bob {

// TODO: add your solution here
std::string hey(std::string msg) {
    // TODO: more semantic, less ad hoc, O(n) ≻? O(2n)
    bool isyell = false;
    bool issilence = true;
    bool isquest = false;
    for (int i = 0; i < (int) msg.length(); i++) {
        //if (!isalpha(msg[i])) {isquest = msg[i] == '?';}
        if (!isspace(msg[i])) {
            issilence = false;
            if (isalpha(msg[i])) {
                isyell = (isupper(msg[i]));
                if (!isyell) {break;}
            }
        }
    }
    // this is a copout
    for (int i = (int) msg.length(); i >= 0; i--) {
        if (isalpha(msg[i])) {break;}
        if (!isspace(msg[i])) {
            isquest = msg[i] == '?';
            if (isquest) {break;}
        }
    }
    if (issilence) {return "Fine. Be that way!";}
    //if (msg[msg.length()-1] == '?') {
    if (isquest) {
        return (isyell) ? "Calm down, I know what I'm doing!" : "Sure.";
    } else {
        return (isyell) ? "Whoa, chill out!" : "Whatever.";
    }
}

}  // namespace bob
